support Ed25519 sign/verify scheme

sa-kib opened the pull request:
This PR adds Ed25519 support as defined in GlobalPlatform TEE Internal Core API v1.3, using libtomcrypt as the backend crypto implementation.
Companion PR for optee_test: [#610](https://github.com/OP-TEE/optee_test/pull/610)

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jenswi-linaro commented on lib/libutee/include/tee_api_defines.h:
```diff
@@ -192,6 +192,7 @@
 #define TEE_ALG_ECDSA_P256                      0x70003041
 #define TEE_ALG_ECDSA_P384                      0x70004041
 #define TEE_ALG_ECDSA_P521                      0x70005041
+#define TEE_ALG_ED25519                         0x70006043
```
Please add a comment at the end of the define that this is from the v1.3 or v1.3.1 spec.

jenswi-linaro replied:
Same for the other new defines in this commit.

sa-kib replied:
done

jenswi-linaro reviewed (commented):
Comment for the commit "libutee: add Ed25519 support"

In the description of the commit. Which version 1.3 or 1.3.1? I guess we would prefer version 1.3.1 as that's the latest.

jenswi-linaro commented on core/include/crypto/crypto.h:
```diff
@@ -209,6 +216,22 @@ TEE_Result crypto_acipher_gen_dh_key(struct dh_keypair *key, struct bignum *q,
+TEE_Result crypto_acipher_ed25519ctx_sign(struct ed25519_keypair *key,
+					  const uint8_t *msg, size_t msg_len,
+					  uint8_t *sig, size_t *sig_len,
+					  uint8_t flag, const uint8_t *ctx);
```
`s/uint8_t flag/bool with_ph/` or something like that. Same for `crypto_acipher_ed25519ctx_verify()`.

sa-kib replied:
I've put `ph_flag` -- just to be consistent with the rest of the code

jenswi-linaro commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2095,6 +2138,114 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+{
+	size_t n;
+	size_t ctx_len;
+	uint8_t ctx[256] = {0};
```
This might be a bit much to keep on the stack. This could be allocated using `mempool_alloc(mempool_default, ...)` instead, pretty much guaranteed to succeed. If the allocation fails there's an error in the mempool configuration.

sa-kib replied:
sure, then I shall use `mempool_calloc(..)`

jenswi-linaro reviewed (commented):
Comments for the commit "core: crypto: add Ed25519 support"

jenswi-linaro reviewed (commented):
Regarding the commit "core: libtomcrypt: add ed25519ctx and ed25519ph support"

Do you have any plans to upstream this?

sjaeckel commented:
c.f. https://github.com/libtom/libtomcrypt/pull/597

larperaxis commented on core/crypto.mk:
```diff
@@ -192,6 +193,7 @@ _CFG_CORE_LTC_XTS := $(CFG_CRYPTO_XTS)
 _CFG_CORE_LTC_CCM := $(CFG_CRYPTO_CCM)
 _CFG_CORE_LTC_AES_DESC := $(call cfg-one-enabled, CFG_CRYPTO_XTS CFG_CRYPTO_CCM)
 $(call force,CFG_CRYPTO_X25519,n,not supported by mbedtls)
+$(call force,CFG_CRYPTO_ED25519,n,not supported by mbedtls)
```
There is already plumbing in crypto.mk to include parts of tomcrypt to get full cipher support when using mbedtls. I would prefer that 25519 is available regardless of the chosen crypto library.

sa-kib replied:
@larperaxis @jforissier then I shall make a separate PR for this, thank you for taking a look

sa-kib commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2095,6 +2138,114 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+	if (res != TEE_SUCCESS)
+		return res;
+
+	tee_ed25519_key = (struct ed25519_keypair *)o->attr;
```
done

sa-kib commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2095,6 +2138,114 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+				 uint32_t param_count)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct ed25519_keypair *tee_ed25519_key = NULL;
```
indeed, fixed

sa-kib commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2095,6 +2138,114 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+	uint8_t ph_flag = 0;
+	uint8_t cx_flag = 0;
+
+	for (n = 0u; n < num_params; n++) {
```
fixed

sa-kib commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2095,6 +2138,114 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+	size_t n;
+	size_t ctx_len;
+	uint8_t ctx[256] = {0};
+	uint8_t ph_flag = 0;
```
fixed

sa-kib commented:
> Comment for the commit "libutee: add Ed25519 support"
> 
> In the description of the commit. Which version 1.3 or 1.3.1? I guess we would prefer version 1.3.1 as that's the latest.

hi @jenswi-linaro , thanks for a review.
Indeed it's a v1.3.1, I've put corresponding references in new declarations and also changed commit messsage.

jenswi-linaro commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2175,22 +2176,24 @@ tee_svc_obj_ed25519_sign(struct ed25519_keypair *key,
 			if (ctx_len > 255)
 				return TEE_ERROR_BAD_PARAMETERS;
-
+			ctx = mempool_calloc(mempool_default, 1, 256);
```
This must be freed when we're done with it.

sa-kib replied:
fixed by [9d501f0](https://github.com/OP-TEE/optee_os/pull/5486/commits/9d501f03af51c0623d67826dd3c43cfcfc4a864d), along with other fixes that I've missed before.

sjaeckel commented on core/include/crypto/crypto.h:
```diff
@@ -209,6 +216,22 @@ TEE_Result crypto_acipher_gen_dh_key(struct dh_keypair *key, struct bignum *q,
+TEE_Result crypto_acipher_ed25519ctx_verify(struct ed25519_keypair *key,
+					    const uint8_t *msg, size_t msg_len,
+					    const uint8_t *sig, size_t sig_len,
+					    bool ph_flag, const uint8_t *ctx);
```
This misses a `ctx_len`

sa-kib replied:
fixed in [30f62e2](https://github.com/OP-TEE/optee_os/pull/5486/commits/30f62e228387757e44ccffdcaf7c37db3a28c602), thank you

jforissier commented on core/include/crypto/crypto.h:
```diff
@@ -172,6 +172,11 @@ struct x25519_keypair {
 
+struct ed25519_keypair {
+	uint8_t *priv;	/* Private value */
+	uint8_t *pub;	/* Public value */
```
Comments not needed, the names are obvious

sa-kib replied:
fixed

jforissier reviewed (commented):
Hi,

- Two very minor comments below. Please also check the CI checkpatch errors, the alignment issue should be fixed.
- I'd like the fixup commits to be squashed if @jenswi-linaro is OK with that. I did a `git rebase -i --autosquash` and there is an issue with the number of arguments to `ed25519ph_sign()` it seems.
- For commit "libutee: add Ed25519 support":
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jenswi-linaro commented:
Yes, please squash the fixup commits.

sa-kib commented on core/crypto.mk:
```diff
@@ -47,8 +47,9 @@ CFG_CRYPTO_ECC ?= y
 CFG_CRYPTO_SM2_DSA ?= y
 CFG_CRYPTO_SM2_KEP ?= y
-# X25519 is only supported by libtomcrypt
+# X25519 and ed25519 are only supported by libtomcrypt
```
fixed

sa-kib commented:
> Hi,
> 
hi @jforissier, thanks for review!

> * Two very minor comments below. Please also check the CI checkpatch errors, the alignment issue should be fixed.

fixed

> * I'd like the fixup commits to be squashed if @jenswi-linaro is OK with that. I did a `git rebase -i --autosquash` and there is an issue with the number of arguments to `ed25519ph_sign()` it seems.

done swuashing & rebasing, but what's the issue with `ed25519ph_sign()` ?

jforissier commented:
> what's the issue with `ed25519ph_sign()` ?

It is introduced with 7 arguments in patch number 2, then called with only 6 in patch number 3, then the final patch adds the missing argument (`cxtlen`).

Please change the subject of patch 2 from "core: libtomcrypt: add ed25519ctx and ed25519ph support" to "libtomcrypt: add ed25519ctx and ed25519ph support".

sa-kib commented:
> > what's the issue with `ed25519ph_sign()` ?
> 
> It is introduced with 7 arguments in patch number 2, then called with only 6 in patch number 3, then the final patch adds the missing argument (`cxtlen`).
>

oh, I see, it's fixed already then
 
> Please change the subject of patch 2 from "core: libtomcrypt: add ed25519ctx and ed25519ph support" to "libtomcrypt: add ed25519ctx and ed25519ph support".

yes, done

jforissier commented:
> oh, I see, it's fixed already then

No. The functions are introduced in patch 2 with a wrong signature, then modified in patch 3. I don't want that, please do the right thing the first time. Unless I'm missing something...

sa-kib commented:
> > oh, I see, it's fixed already then
> 
> No. The functions are introduced in patch 2 with a wrong signature, then modified in patch 3. I don't want that, please do the right thing the first time. Unless I'm missing something...

my bad, I've messed with one of fixups.
Now I've done a proper fixup, squashed, rebased & verified.

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2095,6 +2139,134 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+			   const TEE_Attribute *params, size_t num_params)
+{
+	TEE_Result err;
+	size_t n;
```
ditto

sa-kib replied:
fixed in [785b4be](https://github.com/OP-TEE/optee_os/pull/5486/commits/785b4be2ede4583aa7e251851bdb6ed7905ccbc3)

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2095,6 +2139,134 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+			ctx_len = params[n].content.ref.length;
+			if (ctx_len > 255)
+				return TEE_ERROR_BAD_PARAMETERS;
+			ctx = mempool_calloc(mempool_default, 1, ctx_len + 1);
```
check `ctx != NULL`

sa-kib replied:
fixed in [785b4be](https://github.com/OP-TEE/optee_os/pull/5486/commits/785b4be2ede4583aa7e251851bdb6ed7905ccbc3)

varder commented on core/lib/libtomcrypt/src/pk/ec25519/ec25519_crypto_ctx.c:
```diff
@@ -0,0 +1,40 @@
+  XMEMCPY(buf, &ctxlen8, 1);
+  buf++;
+
+  if (ctxlen > 0u) {
```
In this case the test is needed because the size
of the context is calculated and appends to the context prefix. 
Algorithm tests will not pass in the libtomcrypt

sa-kib replied:
indeed, fixed in [4d7340e](https://github.com/OP-TEE/optee_os/pull/5486/commits/4d7340ef909f3fd98a1dc02c0975e1641178a4ba)

jenswi-linaro commented on core/lib/libtomcrypt/src/pk/ec25519/ec25519_crypto_ctx.c:
```diff
@@ -0,0 +1,40 @@
+/* LibTomCrypt, modular cryptographic library -- Tom St Denis */
+/* SPDX-License-Identifier: Unlicense */
```
It seems that LTC is using this license consistently.

sa-kib replied:
@sjaeckel created that file & code, so I didn't dare to change it.

sa-kib commented on core/lib/libtomcrypt/src/headers/tomcrypt_pk.h:
```diff
@@ -368,10 +368,28 @@ int ed25519_import_pkcs8(const unsigned char *in, unsigned long inlen,
+                   const curve25519_key *private_key);
+int ed25519_verify(const unsigned char *msg, unsigned long msglen,
+                   const unsigned char *sig, unsigned long siglen,
+                         int *stat,
```
yep, I've just tried to retain LTC coding style there.
But I can do it differently ofc, this header seems to have slightly different identation styles mixed.

sa-kib commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2095,6 +2139,134 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+			break;
+
+		case TEE_ATTR_EDDSA_CTX:
+			/* only first provided context if effective */
```
Specification does not explicitly forbid passing several context strings, the decision what to do in this case seems to be left up to implementation. Yet the case with several context strings does look like a clear programming error, so it probably should panic indeed, I'll do a change.

sa-kib replied:
done in [9f65a72](https://github.com/OP-TEE/optee_os/pull/5486/commits/9f65a72a4f85537ad3610cf47d0f83b4db249acc) -- only one context is permitted, also disposed of boilerplate code.

sa-kib commented on core/lib/libtomcrypt/src/pk/ec25519/tweetnacl.c:
```diff
@@ -242,6 +242,27 @@ static int tweetnacl_crypto_hash(u8 *out,const u8 *m,u64 n)
+{
+  unsigned long len;
+  int err;
+  u8 buf[512];
```
in tweetnacl they do all allocations on stack, so we followed their a style (though other allocations in this library are considerably smaller).

As for 512 size considerations are:
255 (context size) + 32 context prefix + 1 flag + 1 ctx len
PREFIX=b"SigEd25519 no Ed25519 collisions" so minimal allocation of 289 seems to be required.
But there are another algorithms that might require context prefix. So it was rounded up to 512.

sjaeckel replied:
@etienne-lms I also thought it could be worth it, but apparently tweetnacl itself is already very stack-heavy.

Each invocation of `tweetnacl_crypto_hash[_ctx]()` inside tweetnacl originates from a function that also calls other functions. Those other functions require way more stack than 512 bytes, ergo that stack size is already required.

If I counted correctly the call-stack of `tweetnacl_crypto_sign(): scalarbase() -> scalarmult() -> add() -> M() -> car25519()` requires those stack variables:

```
typedef i64 gf[16];
  int i; //4 or 8
  i64 c; //8
  i64 i,j,t[31]; //264
  gf a,b,c,d,t,e,f,g,h; //1152
  u8 b; //1
  int i; //4 or 8
  gf q[4]; //512
// max. sum of the above: 1953 bytes
```

sjaeckel replied:
I still liked the idea so there's the last commit of https://github.com/libtom/libtomcrypt/pull/598

sa-kib replied:
Patch backported as a fixup: [ff58986](https://github.com/OP-TEE/optee_os/pull/5486/commits/ff58986367bc764f3c51e78ab8c60a0e4920fa83)

sa-kib commented on core/lib/libtomcrypt/src/pk/ec25519/tweetnacl.c:
```diff
@@ -242,6 +242,27 @@ static int tweetnacl_crypto_hash(u8 *out,const u8 *m,u64 n)
+  XMEMCPY(buf+cs,m,n);
+
+  err = tweetnacl_crypto_hash(out,buf,len);
+  zeromem(buf, len);
```
agree, done in [4d7340e](https://github.com/OP-TEE/optee_os/pull/5486/commits/4d7340ef909f3fd98a1dc02c0975e1641178a4ba)

sa-kib commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2095,6 +2139,134 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+			if (cx_flag)
+				break;
+			ctx_len = params[n].content.ref.length;
+			if (ctx_len > 255)
```
done in [785b4be](https://github.com/OP-TEE/optee_os/pull/5486/commits/785b4be2ede4583aa7e251851bdb6ed7905ccbc3)

sa-kib commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2095,6 +2139,134 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+			if (!ctx)
+				return TEE_ERROR_OUT_OF_MEMORY;
+			memcpy(ctx, params[n].content.ref.buffer, ctx_len);
+			ctx[ctx_len] = 0;
```
fixed in [785b4be](https://github.com/OP-TEE/optee_os/pull/5486/commits/785b4be2ede4583aa7e251851bdb6ed7905ccbc3)

sa-kib commented on core/lib/libtomcrypt/src/pk/ec25519/ec25519_crypto_ctx.c:
```diff
@@ -0,0 +1,40 @@
+  const unsigned long prefix_len = strlen(prefix);
+  const unsigned char ctxlen8 = (unsigned char)ctxlen;
+
+  if (ctxlen > 255u) return CRYPT_INPUT_TOO_LONG;
```
255 is the maximum length of context string according to RFC.
If the number here is too magic -- can we get away with a separate macro, e.g. `LTC_ED25519_CTX_MAX_LENGTH`?

sa-kib commented on core/lib/libtomcrypt/src/pk/ec25519/tweetnacl.c:
```diff
@@ -451,23 +472,22 @@ static int unpackneg(gf r[4],const u8 p[32])
 
   *stat = 0;
   if (*mlen < smlen) return CRYPT_BUFFER_OVERFLOW;
-  *mlen = -1;
```
agree, done in [c7a6adf](https://github.com/OP-TEE/optee_os/pull/5486/commits/c7a6adf5f79d8b76d54634755dbc4b05c175a8de)

sa-kib commented on core/lib/libtomcrypt/src/pk/ed25519/ed25519_verify.c:
```diff
@@ -55,16 +48,99 @@ int ed25519_verify(const  unsigned char *msg, unsigned long msglen,
 
 #ifdef LTC_CLEAN_STACK
-   zeromem(m, mlen);
+   zeromem(m, msglen + siglen);
```
agree, removed in [c7a6adf](https://github.com/OP-TEE/optee_os/pull/5486/commits/c7a6adf5f79d8b76d54634755dbc4b05c175a8de)

sa-kib replied:
made a separate commit with a better description regarding its origin: [2824954](https://github.com/OP-TEE/optee_os/pull/5486/commits/2824954dfc60ac9acd1ad8ac3443fdf47638c918)

sa-kib commented on core/lib/libtomcrypt/src/pk/ed25519/ed25519_sign.c:
```diff
@@ -64,6 +58,83 @@ int ed25519_sign(const unsigned char  *msg, unsigned long msglen,
+                    const curve25519_key *private_key)
+{
+   int err;
+   unsigned char ctx_prefix[512] = {0};
```
done in [3f5a052](https://github.com/OP-TEE/optee_os/pull/5486/commits/3f5a05234885f64a7877d2b0f14294eba8e97526)

sa-kib commented on core/crypto/crypto.c:
```diff
@@ -825,6 +825,65 @@ TEE_Result crypto_acipher_x25519_shared_secret(struct x25519_keypair
+{
+	return TEE_ERROR_NOT_IMPLEMENTED;
+}
+
```
removed in [a850e12](https://github.com/OP-TEE/optee_os/pull/5486/commits/a850e12a5bd4633c6b8d9d12c1acd73aa0c49c35)

sa-kib commented on core/lib/libtomcrypt/ed25519.c:
```diff
@@ -0,0 +1,175 @@
+	if (!key)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	memset(key, 0, sizeof(*key));
```
it's even nicer w/o this variable, refactored in [a850e12](https://github.com/OP-TEE/optee_os/pull/5486/commits/a850e12a5bd4633c6b8d9d12c1acd73aa0c49c35)

sa-kib commented on core/lib/libtomcrypt/ed25519.c:
```diff
@@ -0,0 +1,175 @@
+	if (!key)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	memcpy(private_key.priv, key->priv, 32);
```
agree, fixed in [a850e12](https://github.com/OP-TEE/optee_os/pull/5486/commits/a850e12a5bd4633c6b8d9d12c1acd73aa0c49c35)

sa-kib commented on core/lib/libtomcrypt/ed25519.c:
```diff
@@ -0,0 +1,175 @@
+	memcpy(private_key.priv, key->priv, 32);
+	memcpy(private_key.pub, key->pub, 32);
+
+	if (ed25519_sign(msg, msg_len, sig, (unsigned long *)sig_len,
```
agree, we can do w/o a cast: [a850e12](https://github.com/OP-TEE/optee_os/pull/5486/commits/a850e12a5bd4633c6b8d9d12c1acd73aa0c49c35)

sa-kib replied:
my bad, we do need a cast. Made a separate variable, as suggested: [332157e](https://github.com/OP-TEE/optee_os/pull/5486/commits/332157e7e07a85513c04d9157810bf844d5c31d5)

jforissier replied:
> my bad, we do need a cast. Made a separate variable, as suggested: [332157e](https://github.com/OP-TEE/optee_os/pull/5486/commits/332157e7e07a85513c04d9157810bf844d5c31d5)

The intermediate variable has to be an `unsigned long` as Etienne mentioned, not a pointer.

sa-kib replied:
please see [a051fe3](https://github.com/OP-TEE/optee_os/pull/5486/commits/a051fe3767819768e74974ea1a880a161c0abd51), have I got it right?

jforissier replied:
Absolutely, thanks.

sa-kib commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2095,6 +2139,134 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+			 const TEE_Attribute *params, size_t num_params)
+{
+	TEE_Result err;
+	size_t n;
```
done in [15055f0](https://github.com/OP-TEE/optee_os/pull/5486/commits/15055f043c6e61c4114e05691f674c6975594865)

jforissier commented:
Please do not use the "core:" prefix for pure LibTomCrypt changes. I know LTC is used only in the TEE core not as a user space library, but I prefer to reserve "core: libtomcrypt: ..." for the OP-TEE specific adaptation layer and "libtomcrypt: ..." for the the rest. Makes it a bit easier to see which commit touches what.

Overall this looks good to me, I will review again once fixups are squashed, @etienne-lms let us now when you want them squashed.

etienne-lms commented on core/lib/libtomcrypt/ed25519.c:
```diff
@@ -0,0 +1,175 @@
+				    ctx, ctxlen, &private_key) != CRYPT_OK)
+			return TEE_ERROR_BAD_PARAMETERS;
+	}
+
```
Public key content is not that sensible. I think it's not worth wiping its content.

etienne-lms reviewed (commented):
few minor comments.
@sa-kib, you can squash the fixup commits.

sa-kib commented on core/include/crypto/crypto.h:
```diff
@@ -172,6 +172,11 @@ struct x25519_keypair {
 
+struct ed25519_keypair {
+	uint8_t *priv;
+	uint8_t *pub;
```
it was [agreed](https://github.com/OP-TEE/optee_os/pull/5486#discussion_r952366948) with @jforissier to be unnecessary.

etienne-lms replied:
ok

jforissier commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2095,6 +2141,125 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+				break;
+
+			*ctx = params[n].content.ref.buffer;
+			if (*ctx == NULL)
```
`if (!*ctx)`

sa-kib replied:
done

jforissier commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2095,6 +2141,125 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+			if (*ctx_len > TEE_ED25519_CTX_MAX_LENGTH)
+				return TEE_ERROR_BAD_PARAMETERS;
+
+			if (*ctx_len == 0)
```
`if (!*ctx_len)`

sa-kib replied:
done

jforissier reviewed (commented):
With the below minor comments addressed (and please also fix the long line in the commit description spotted by checkpatch):

"libtomcrypt: fix excessive memory clean"
"libtomcrypt: add ed25519ctx and ed25519ph support"
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

"core: libtomcrypt: add Ed25519 support"
"core: crypto: add Ed25519 support "
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

sa-kib commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2095,6 +2141,125 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+
+	*ctx = NULL;
+
+	for (; n < num_params; n++) {
```
done

sa-kib commented:
squashed & rebased. Also get rid of `mempool_alloc()`, as we don't modify context and to make thinks a bit simpler.

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commit 
"libtomcrypt: fix excessive memory clean"

`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
"libutee: add Ed25519 support"
"libtomcrypt: add ed25519ctx and ed25519ph support"
"core: libtomcrypt: add Ed25519 support"
"core: crypto: add Ed25519 support"

sjaeckel commented:
Maybe you want to update your libtomcrypt? c.f. https://github.com/libtom/libtomcrypt/pull/599

jforissier commented:
> Maybe you want to update your libtomcrypt? c.f. [libtom/libtomcrypt#599](https://github.com/libtom/libtomcrypt/pull/599)

That would be a good thing to do indeed. The last full sync dates back to 4 years ago,  although we have picked a few fixes from upstream since then. It doesn't necessarily have to be done before this PR is merged however.

@sjaeckel can we consider that the current "develop" branch is stable?

sjaeckel commented:
> @sjaeckel can we consider that the current "develop" branch is stable?

yes, I'd say current develop is pretty stable.

The only things happening sometime would be: 
1. what will happen with the ECC API after https://github.com/libtom/libtomcrypt/pull/450 and https://github.com/libtom/libtomcrypt/pull/477 are tackled... but I don't expect this to be really big changes, for 477 I've even a compatible way in mind.
2. whether we will follow up on https://github.com/libtom/libtomcrypt/pull/515 and the other registries

sa-kib commented:
hi @jenswi-linaro, can this PR be considered in fit for inclusion?

jenswi-linaro commented:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
@sa-kib sorry for the delay on this PR. On second thought, I'd like to have it merged **after** LibTopmCrypt is updated in the master branch (that is #5539 which will be squash-merged into master). In the meantime, I have tried re-basing your patches onto the new LTC (without the two `libtomcrypt:` patches of course, which are already included upstream) and I there were only minor conflicts/issues. You will need to apply the below diff to "core: libtomcrypt: add Ed25519 support" and all should be well.

```diff
diff --git a/core/lib/libtomcrypt/ed25519.c b/core/lib/libtomcrypt/ed25519.c
index 33c1d852..b85629b8 100644
--- a/core/lib/libtomcrypt/ed25519.c
+++ b/core/lib/libtomcrypt/ed25519.c
@@ -7,6 +7,7 @@
 #include <crypto/crypto.h>
 #include <stdlib.h>
 #include <string.h>
+#include <string_ext.h>
 #include <tee_api_types.h>
 #include <trace.h>
 #include <utee_defines.h>
@@ -65,7 +66,7 @@ TEE_Result crypto_acipher_ed25519_sign(struct ed25519_keypair *key,
 	unsigned long siglen;
 	curve25519_key private_key = {
 		.type = PK_PRIVATE,
-		.algo = PKA_ED25519,
+		.algo = LTC_OID_ED25519,
 	};
 
 	if (!key)
@@ -94,7 +95,7 @@ TEE_Result crypto_acipher_ed25519ctx_sign(struct ed25519_keypair *key,
 	unsigned long siglen;
 	curve25519_key private_key = {
 		.type = PK_PRIVATE,
-		.algo = PKA_ED25519,
+		.algo = LTC_OID_ED25519,
 	};
 
 	if (!key)
@@ -126,7 +127,7 @@ TEE_Result crypto_acipher_ed25519_verify(struct ed25519_keypair *key,
 	int stat = 0;
 	curve25519_key public_key = {
 		.type = PK_PUBLIC,
-		.algo = PKA_ED25519,
+		.algo = LTC_OID_ED25519,
 	};
 
 	if (!key)
@@ -153,7 +154,7 @@ TEE_Result crypto_acipher_ed25519ctx_verify(struct ed25519_keypair *key,
 	int stat = 0;
 	curve25519_key public_key = {
 		.type = PK_PUBLIC,
-		.algo = PKA_ED25519,
+		.algo = LTC_OID_ED25519,
 	};
 
 	if (!key)
diff --git a/core/lib/libtomcrypt/src/pk/ec25519/sub.mk b/core/lib/libtomcrypt/src/pk/ec25519/sub.mk
index e3d07c0e..5040c391 100644
--- a/core/lib/libtomcrypt/src/pk/ec25519/sub.mk
+++ b/core/lib/libtomcrypt/src/pk/ec25519/sub.mk
@@ -1,2 +1,3 @@
+srcs-y += ec25519_crypto_ctx.c
 srcs-y += ec25519_export.c
 srcs-y += tweetnacl.c
diff --git a/core/lib/libtomcrypt/src/pk/ed25519/sub.mk b/core/lib/libtomcrypt/src/pk/ed25519/sub.mk
index bdd41059..cb2206da 100644
--- a/core/lib/libtomcrypt/src/pk/ed25519/sub.mk
+++ b/core/lib/libtomcrypt/src/pk/ed25519/sub.mk
@@ -3,6 +3,5 @@ srcs-y += ed25519_import.c
 srcs-y += ed25519_import_pkcs8.c
 srcs-y += ed25519_import_x509.c
 srcs-y += ed25519_make_key.c
-srcs-y += ed25519_set_key.c
 srcs-y += ed25519_sign.c
 srcs-y += ed25519_verify.c
```

jforissier commented:
@sa-kib @varder I have now merged the LibTomCrypt update, so could you please rebase this PR on top of master and apply the Ack/Review tags given above? Thanks!

jforissier commented:
@sa-kib @varder thanks for this contribution and for working with upstream LibTomCrypt too. I am merging this despite the IBART failure because I know it is caused by the missing commit b7563bab0 in your branch.
